Kernel/Ptrace.cpp

```diff
@@ -0,0 +1,165 @@
+#include <Kernel/Process.h>
+#include <Kernel/Ptrace.h>
+#include <Kernel/Thread.h>
+#include <Kernel/ThreadTracer.h>
+#include <Kernel/VM/MemoryManager.h>
+#include <Kernel/VM/ProcessPagingScope.h>
+
+namespace Ptrace {
+
+KResultOr<u32> handle_syscall(const Kernel::Syscall::SC_ptrace_params& params, Process& caller)
+{
+    if (params.request == PT_TRACE_ME) {
+        if (Thread::current->tracer())
+            return KResult(-EBUSY);
+
+        caller.set_wait_for_tracer_at_next_execve(true);
+        return KSuccess;
+    }
+
+    if (params.pid == caller.pid())
+        return KResult(-EINVAL);
+
+    Thread* peer = nullptr;
+    {
+        InterruptDisabler disabler;
+        peer = Thread::from_tid(params.pid);
+    }
+    if (!peer)
+        return KResult(-ESRCH);
+
+    if (peer->process().uid() != caller.euid())
```

Comment:
Thinking about it, we should probably also disallow tracing `setuid` processes.
PT_POKE & PT_SETREGS could be used in a crafty way for privilege escalation!